### DISCUSSION TITLE
path problem, how can i solve this ?

### DISCUSSION THREAD

**Discussion by StefanSa** [vectordotdev/vector#10739] [Q&A] [answered]
Hi there,
have this:
```
[transforms.ecs_url_mapping_atp]
inputs = ["ecs_src_mapping_atp"]
type   = "remap"
source = '''
if exists(.sophos.xg.url) {
   .url.original = del(.sophos.xg.url)
}
. = parse_url!(.url.original)
.url.domain = del(.host)
.url.path = del(.path)
.url.scheme = del(.scheme)
.url.fragment = del(.fragment)
.url.password = del(.password)
.url.username = del(.username)
.url.port = del(.xg.port)
.url.query = del(.query)

[transforms.geo_mapping_src_atp]
type = "geoip"
inputs = ["ecs_url_mapping_atp"]
database = "/etc/vector/GeoIP/GeoLite2-City.mmdb"
source = "source.ip"
target = "source.geo"
```
With this construct, vector aborts with the error message that it cannot find the field for `.source.ip` (geoip)
and that the values of `parce_url` are make complete the `json` value defect.

i have understood that this is a `path` problem, for example, if i set the path as in the master pipeline,
`. = set!(., path: ["sophos", "xg"], data: parse_url!(.url.original))`
the previous events with this path will be deleted / missing.
However the value of the `parse_url` is correct and also the `json` result.
The question is, what is the correct value of the `path` for `parce_url` function?

**Comment by StefanSa** [comment]
Ok. took another path like the one in the master pipeline, so it seems to work.
Apparently a path must be passed here.
The question is how to do it right.

**Comment by spencergilbert** [comment, marked as answer]
Can you share an example event thats coming into that `remap` component, as well as what you'd expect the event to look like after?

One thing that I notice immediately about the script is this:

```coffee
if exists(.sophos.xg.url) {
   .url.original = del(.sophos.xg.url)
}
# Everything below assumes the above `if` statement is true,
# which suggests everything below should be included in the
# above block. Otherwise you could be calling these functions
# on fields that don't exist and erroring.
. = parse_url!(.url.original)
.url.domain = del(.host)
.url.path = del(.path)
.url.scheme = del(.scheme)
.url.fragment = del(.fragment)
.url.password = del(.password)
.url.username = del(.username)
.url.port = del(.xg.port)
.url.query = del(.query)
```

That is to say, the following seems more correct and robust:

```coffee
if exists(.sophos.xg.url) {
  .url.original = del(.sophos.xg.url)
  . = parse_url!(.url.original)
  .url.domain = del(.host)
  .url.path = del(.path)
  .url.scheme = del(.scheme)
  .url.fragment = del(.fragment)
  .url.password = del(.password)
  .url.username = del(.username)
  .url.port = del(.xg.port)
  .url.query = del(.query)
}
```

**Reply by StefanSa** [reply]
Hi @spencergilbert Spencer,

Incoming value
`.sophos.xg.url = http://sophostest.com/callhom e/index.html`

Expected result:
```
  "url": {
    "domain": "sophostest.com",
    "original": "http://sophostest.com/callhom e/index.html",
    "path": "/callhome/index.html",
    "scheme": "http"
  }
```

FYI, as mentioned.
If i take any path here, but not the one from the master pipeline, everything is OK.
```
. = set!(., path: ["s", "g"], data: parse_url!(.url.original))
.url.domain = del(.s.g.host)
.url.path = del(.s.g.path)
.url.scheme = del(.s.g.scheme)
.url.fragment = del(.s.g.fragment)
.url.password = del(.s.g.password)
.url.username = del(.s.g.username)
.url.port = del(.s.g.port)
.url.query = del(.s.g.query)
```

**Reply by spencergilbert** [reply]
Is it possible the path is actually `."sophos.xg".url` and not `.sophos.xg.url`?

```json
{
  "sophos.xg": {
    "url": ""
  }
}
```

vs

```json
{
  "sophos.xg.url": ""
}
```

**Reply by StefanSa** [reply]
@spencergilbert 
no Spencer, i have modified the path you provided to follow the `elastic ecs scheme`. 
I currently use this (without problems). 
` . = set!(., path: ["sophos", "xg"], data: parse_key_value!(.message))`

**Reply by spencergilbert** [reply]
OH - @StefanSa perhaps I see the problem, `. = parse_url!(.url.original)` will _replace_ the entire event with the parsed URL. You may be wanting to do this instead:

```coffee
if exists(.sophos.xg.url) {
  .url.original = del(.sophos.xg.url)
  . |= parse_url!(.url.original) # `|=` is the "assignment operator"
  # ref: https://vector.dev/docs/reference/vrl/expressions/#assignment-example-object-merge-assignment
  .url.domain = del(.host)
  .url.path = del(.path)
  .url.scheme = del(.scheme)
  .url.fragment = del(.fragment)
  .url.password = del(.password)
  .url.username = del(.username)
  .url.port = del(.xg.port)
  .url.query = del(.query)
}
```

**Comment by StefanSa** [comment]
@spencergilbert Spencer
here is an example with successful parsing with Vector :)
[example.json.zip](https://github.com/vectordotdev/vector/files/7823432/example.json.zip)

**Reply by spencergilbert** [reply]
This looks really nice! I'd love to see your full config once you have it finished, I spent some time a year ago trying to setup a config to transform to ECS and I eventually gave up 😆

**Reply by StefanSa** [reply]
@spencergilbert 
:+1: Thank you, perfect.
something about the background.
i am the initial maintainer of some firewall modules from elastic filebeat, e.g. sophos-xg and juniper. I now want to realize these modules, plus others, with vector. The goal should be to get along without the ELK stack and to realize this using only vector and opensearch.
